docs/proposals/external-lb-source-ip-preservation.md

```diff
@@ -0,0 +1,226 @@
+To steer traffic only onto nodes that have endpoints for the service, we program the LB to perform
+node healthchecks. The kube-proxy daemons running on each node will be responsible for responding
+to these healthcheck requests from the cloud provider LB healthchecker.
+kube-proxy already watches for Service and Endpoint changes, it will maintain an in-memory lookup
+table indicating the number of local endpoints for each service.
+For a value of zero, it responds with a health check failure, and success for non-zero values.
+
+Healthchecks are programmable with a min period of 1 second on most cloud provider LBs, and min
+failures to trigger node health state change can be configurable from 2 through 5.
+
+This will allow much faster transition times on the order of 1-5 seconds, and involve no
+API calls to the cloud provider (and hence reduce the impact of API unreliability), keeping the
+time window where traffic might get directed to nodes with no local endpoints to a minimum.
+
+## API Changes
+
+### Local Endpoint Recognition Support
+
+To allow kube-proxy to recognize if an endpoint is local requires that the EndpointAddress struct
+should also contain the NodeName it resides on. This new string field will be read-only and 
+populated *only* by the Endpoints Controller.
+
+### Service Annotation to opt-in for new behaviour
+
+A new annotation ```service.alpha.kubernetes.io/cloud-lb-only-node-local-endpoints``` will be recognized
+by the service controller. Services that wish to opt-in to the new LoadBalancer behaviour must
+annotate the Service to request the new ESIPP behaviour.
+
+### NodePort allocation for HealthChecks
+
+An additional nodePort allocation will be necessary for services that are of type LoadBalancer and
+have the new annotation specified. This additional nodePort is necessary for kube-proxy to listen for
+healthcheck requests on all nodes.
+
+## Cloud Provider support
+
+This feature is added as an opt-in annotation.
+Default behaviour of LoadBalancer type services will be unchanged for all Cloud providers.
+The annotation will be ignored by existing cloud provider libraries until they add support.
+
+### GCE 1.4
+
+For the 1.4 release, this feature will be implemented for the GCE cloud provider.
+
+### AWS TBD
+
+```discuss timelines and feasibility with Kubernetes sig-aws team members```
+
+### Openstack TBD
+
+```discuss timelines and feasibility with Kubernetes Openstack team members```
+
+### Azure TBD
+
+```discuss timelines and feasibility with Kubernetes Azure team members```
+
+## Behavior Changes expected
+
+### Traffic Balancing Changes
+
+GCE/AWS load balancers do not provide weights for their target pools. This was not an issue with the old LB
+kube-proxy rules which would correctly balance across all endpoints.
+
+With the new functionality, the external traffic will not be equally load balanced across pods, but rather
```

Comment:
AWS supports setting up multiple listeners that hit different NodePorts, but their traffic needs to hit different ports on the load balancer's external IP, so, no, it's not going to work.